### PR TITLE
[hmac,dv] Block-level tests with SHA-2 NIST test vectors for new digest sizes 384/512

### PR DESCRIPTION
**Pull request by gdessouky**
This adds block-level testing with SHA-2 NIST test vectors for newer digest sizes 384/512, which addresses https://github.com/lowRISC/opentitan/issues/22932 in part. 
Block-level testing for HMAC with different key lengths is not yet added. Once in, this would close out https://github.com/lowRISC/opentitan/issues/22932.

### CODE REVIEW

File: hw/dv/sv/test_vectors/test_vectors_pkg.sv
Comment (martin-velay) [inline hw/dv/sv/test_vectors/test_vectors_pkg.sv:29]: Why not the HMAC 384/512?

File: hw/dv/sv/test_vectors/test_vectors_pkg.sv
Comment (martin-velay) [inline hw/dv/sv/test_vectors/test_vectors_pkg.sv:150]: Here maybe precise SHA2:
`SHA2-256/SHA2-384/SHA2-512/HMAC/SHA3SHAKE/CSHAKE/KMAC`

File: hw/dv/sv/test_vectors/test_vectors_pkg.sv
Comment (martin-velay) [inline hw/dv/sv/test_vectors/test_vectors_pkg.sv:154]: Same as above: Here maybe precise SHA2:
`SHA2-256/SHA2-384/SHA2-512/HMAC/SHA3SHAKE/CSHAKE/KMAC`

File: hw/ip/hmac/dv/env/seq_lib/hmac_base_vseq.sv
Comment (martin-velay) [inline hw/ip/hmac/dv/env/seq_lib/hmac_base_vseq.sv:53]: Maybe add a TODO here, as Illegal won't be hit and same for the digest size.
Do we need to raise an issue for that?

File: hw/ip/hmac/dv/env/seq_lib/hmac_base_vseq.sv
Comment (martin-velay) [inline hw/ip/hmac/dv/env/seq_lib/hmac_base_vseq.sv:45]: I would prefer to have separated constraints, it will allow us more flexibility if we need to release one for a specific test:

```
  constraint key_length_c {
    $countones(key_length) == 1 dist {
      1 :/ 10,  // Key_128/Key_256/Key_384/Key_512/Key_1024/Key_None
      0 :/ 0    // Illegal -> should get casted to Key_None in HW
    };
  }

  constraint digest_size_c {
    $countones(digest_size) == 1 dist {
      1 :/ 10,  // SHA2_256/SHA2_384/SHA2_512/SHA2_None
      0 :/ 0    // Illegal -> should get casted to SHA2_None in HW
    };
  }

```

File: hw/ip/hmac/dv/env/seq_lib/hmac_test_vectors_hmac_vseq.sv
Comment (martin-velay) [inline hw/ip/hmac/dv/env/seq_lib/hmac_test_vectors_hmac_vseq.sv:5]: As we are changing this comment, can precise from where those vectors are coming as I spent time to understand the purpose of this test. I didn't get at all that those vectors are actually provided by the NIST as a golden reference.

File: hw/ip/hmac/dv/env/seq_lib/hmac_test_vectors_hmac_vseq.sv
Comment (martin-velay) [inline hw/ip/hmac/dv/env/seq_lib/hmac_test_vectors_hmac_vseq.sv:17]: Why commented out? Maybe add a TODO with an issue #?

File: hw/ip/hmac/dv/env/seq_lib/hmac_test_vectors_sha_vseq.sv
Comment (martin-velay) [inline hw/ip/hmac/dv/env/seq_lib/hmac_test_vectors_sha_vseq.sv:5]: Same as above, can we say that those vectors and expected values are provided by the NIST as a golden reference?

File: hw/ip/hmac/dv/env/seq_lib/hmac_test_vectors_sha_vseq.sv
Comment (martin-velay) [inline hw/ip/hmac/dv/env/seq_lib/hmac_test_vectors_sha_vseq.sv:12]: I think `key_length` and `digest_size` are already declared into the `hmac_base_vseq`. So it should be removed from here.

Same for the `key`

File: hw/ip/hmac/dv/env/seq_lib/hmac_test_vectors_sha_vseq.sv
Comment (martin-velay) [inline hw/ip/hmac/dv/env/seq_lib/hmac_test_vectors_sha_vseq.sv:78]: Can you swap `digest_size` with `parsed_vectors[j].exp_digest` tohave the same ordering as above? (and needs to be changed into the `base_seq` as well

File: hw/ip/hmac/dv/env/seq_lib/hmac_test_vectors_sha_vseq.sv
Comment (martin-velay) [inline hw/ip/hmac/dv/env/seq_lib/hmac_test_vectors_sha_vseq.sv:85]: Maybe better to point to the pkg parameters:
`SHA2_256` and `Key_256`

Same for the 2 others below

File: hw/ip/hmac/dv/env/seq_lib/hmac_test_vectors_sha_vseq.sv
Comment (martin-velay) [inline hw/ip/hmac/dv/env/seq_lib/hmac_test_vectors_sha_vseq.sv:87]: Add something like:
```
uvm_info(`gfn, $sformatf("Starting SHA-2/HMAC 256 NIST test vectors..."), UVM_LOW)
```

**Review by martin-velay — APPROVED** [review]
Could be merged once the comments have been taken into account

File: hw/dv/sv/test_vectors/test_vectors_pkg.sv
Comment (gdessouky) [reply hw/dv/sv/test_vectors/test_vectors_pkg.sv:154]: Here it is only the format for SHA-2 256/SHA-2 384/SHA-2 SHA512/HMAC/SHA3 test vectors. CSHAKE and KMAC test vectors format come below.

File: hw/ip/hmac/dv/env/seq_lib/hmac_test_vectors_sha_vseq.sv
Comment (gdessouky) [reply hw/ip/hmac/dv/env/seq_lib/hmac_test_vectors_sha_vseq.sv:78]: does it make a difference? `feed_vectors` is a different function and its parameter is the vector list unparsed, so it is not the same, but I am fine with either order :)

Comment (martin-velay) [reply hw/ip/hmac/dv/env/seq_lib/hmac_test_vectors_sha_vseq.sv:78]: No it's more to keep the same logic and it's something I have in my PR https://github.com/lowRISC/opentitan/pull/23116 as it was causing me some troubles, and I took the order with digest at the end 😁

File: hw/dv/sv/test_vectors/test_vectors_pkg.sv
Comment (martin-velay) [reply hw/dv/sv/test_vectors/test_vectors_pkg.sv:154]: Oh yes, sure it's bad copy/paste from me

File: hw/ip/hmac/dv/env/seq_lib/hmac_test_vectors_sha_vseq.sv
Comment (gdessouky) [reply hw/ip/hmac/dv/env/seq_lib/hmac_test_vectors_sha_vseq.sv:85]: good catch, yes

**Comment by gdessouky** [comment]
Thanks @martin-velay I've pushed again now but a single test run (when it's not smoke) is now very long since it runs through all test vectors for each digest size: SHA-2 256 then 384 then 512, and some time out. I think we should be splitting these up into 3 tests for each digest size and we shouldn't merge now.

File: hw/ip/hmac/dv/env/seq_lib/hmac_base_vseq.sv
Comment (gdessouky) [reply hw/ip/hmac/dv/env/seq_lib/hmac_base_vseq.sv:55]: I tested them at some point in the weeks leading up to M4, to make sure they are correctly cast into `SHA2_None/Key_None` and predicted correctly in the scb, but I'll rework the constraint to hit these and run a block-level regression at my end to make sure.

**Comment by gdessouky** [comment]
I've pushed again to this PR and split the test into 3 tests for each of the digest sizes (I think some timeouts might happen still but we'll see and adjust accordingly), and also released the constraints to hit illegal `digest size/key length`. I've tested/implemented changes for this before to make sure they are casted correctly in the HW and predicted correctly in the scoreboard, but I'm still running a local block-level regression at my end to confirm. Once merged, this should also close https://github.com/lowRISC/opentitan/issues/23592.

**Review by martin-velay — CHANGES_REQUESTED** [review]

File: hw/ip/hmac/dv/env/hmac_env_pkg.sv
Comment (gdessouky) [inline hw/ip/hmac/dv/env/hmac_env_pkg.sv:189]: @martin-velay this is to also return 0 for all invalid input when we release the constraints to include invalid digest sizes/key lengths.

**Comment by gdessouky** [comment]
@martin-velay, thanks for your feedback and comments, I've addressed them now in my last push after rebasing.
@vogelpi, can we go ahead and merge?

File: hw/ip/hmac/dv/env/hmac_env_pkg.sv
Comment (martin-velay) [reply hw/ip/hmac/dv/env/hmac_env_pkg.sv:189]: Yes, I had done this modif but it looks like it has been lost during my rebase somehow. Thanks and sorry!

File: hw/ip/hmac/dv/env/seq_lib/hmac_base_vseq.sv
Comment (martin-velay) [inline hw/ip/hmac/dv/env/seq_lib/hmac_base_vseq.sv:462]: This `digest_index` is not used from what I can see

File: hw/ip/hmac/dv/env/seq_lib/hmac_test_vectors_sha_vseq.sv
Comment (martin-velay) [inline hw/ip/hmac/dv/env/seq_lib/hmac_test_vectors_sha_vseq.sv:26]: Here you should probably add a check like:
```
if (digest_size_arg == "") begin
  `uvm_fatal(`gfn, "This test sequence requires a command-line extra argument: sha2_digest_size=SHA2_256/SHA2_384/SHA2_512"
end
```

**Review by martin-velay — CHANGES_REQUESTED** [review]
There is at least one change which will require a change in case the sequence is misused

File: hw/ip/hmac/dv/env/seq_lib/hmac_base_vseq.sv
Comment (gdessouky) [reply hw/ip/hmac/dv/env/seq_lib/hmac_base_vseq.sv:462]: true, not anymore, thanks for catching.

**Review by martin-velay — APPROVED** [review]
Thanks for the changes, now it LGTM!

**Review by rswarbrick — APPROVED** [review]
This looks sensible to me and I'm pleased to see that @martin-velay (who knows much more about HMAC than I do) agrees. Thanks!

**Comment by gdessouky** [comment]
The failing CI tests are unrelated to the changes in this PR, can we go ahead and merge @rswarbrick?

**Pull request merged**
